### PR TITLE
EZP-24744 - Increase password security

### PR DESCRIPTION
**Pull request by glye**
> https://jira.ez.no/browse/EZP-24744
> Based on https://github.com/netgen/ezpublish-legacy/commit/7d5bd88cb7ceba3e3ec1c349786e1022d375d38f by @emodric 
> which is based on https://github.com/ezsystems/ezpublish-legacy/pull/1195 by @pedroresende 
> Same for new stack: https://github.com/ezsystems/ezpublish-kernel/pull/2095
> Ready for review

Added PASSWORD_HASH_PHP_DEFAULT in addition to PASSWORD_HASH_BCRYPT
Set PASSWORD_HASH_PHP_DEFAULT as default

### CODE REVIEW

**Comment by glye** [comment]
Test failure is not my fault - it cannot download php-5.3.tar.bz2

File: update/database/postgresql/7.0/dbupdate-5.4.0-to-7.0.0.sql [hunk 1]
```diff
@@ -0,0 +1,4 @@
+UPDATE ezsite_data SET value='7.0.0' WHERE name='ezpublish-version';
```
Comment (emodric) [inline update/database/postgresql/7.0/dbupdate-5.4.0-to-7.0.0.sql]: Should this be 6.12?

File: share/db_data.dba [hunk 2]
```diff
@@ -23048,16 +23048,16 @@ keywords=cms, publish, e-commerce, content management, development framework',
         0 => '10',
         1 => 'nospam@ez.no',
         2 => 'anonymous',
-        3 => '4e6f6184135228ccd45f8233d72a0363',
-        4 => '2',
+        3 => '$2y$10$wUrvpYTs9KZ0mtB9Gzr.G.nk6wGhPYzcH3jl9Tk5DtxJN8povoQi2',
+        4 => '6',
```
Comment (emodric) [inline share/db_data.dba]: I'd like to see this consistent with the default value in config (`php_default`)

File: update/database/postgresql/7.0/dbupdate-5.4.0-to-7.0.0.sql (hunk 1 above)
Comment (andrerom) [reply update/database/postgresql/7.0/dbupdate-5.4.0-to-7.0.0.sql]: yes

Comment (glye) [reply update/database/postgresql/7.0/dbupdate-5.4.0-to-7.0.0.sql]: @emodric @andrerom You mean set to `6.12.0` in file `*/6.12/dbupdate-5.4.0-to-6.12.0.sql`, and same for mysql, right?

Comment (emodric) [reply update/database/postgresql/7.0/dbupdate-5.4.0-to-7.0.0.sql]: Yep, exactly.

File: share/db_data.dba (hunk 2 above)
Comment (glye) [reply share/db_data.dba]: @emodric @andrerom Ok. Do you know which password should be used for the anonymous hash? If it matters at all... :)

Comment (emodric) [reply share/db_data.dba]: IIRC, I generated those hashes for `publish` password, for all accounts.

Comment (glye) [reply share/db_data.dba]: Nope, I checked and that's not it. But I can as well use it, I don't think it matters what it is.

Comment (emodric) [reply share/db_data.dba]: You're right. It's `anonymous` for anonymous user and `publish` for admin user.

Comment (glye) [reply share/db_data.dba]: That doesn't fit the existing data either. I used `publish` for both.

**Comment by glye** [comment]
@emodric Changed dbupdate version number, and updated hashes in db data to php_default type.

**Comment by emodric** [comment]
A test is failing now, it seems, so once it's fixed, :+1:

**Comment by glye** [comment]
Hm.. this test is kind of obsolete now, and if we're going to keep it we'll have to keep a db entry with md5_user hash, too.

**Comment by emodric** [comment]
Can we update it so it can use `php_default` instead of `md5_user` ?

**Comment by glye** [comment]
The test makes no sense with php_default, as it doesn't use usernames at all in the hash.
Looking at the code, it should actually work, though. I'll dig some more. 👷

**Comment by glye** [comment]
Reverting db data update, to verify that the problem is indeed there. (Not reproducible locally)

File: composer.json [hunk 3]
```diff
@@ -57,14 +57,15 @@
         "zetacomponents/persistent-object": "~1.8",
         "zetacomponents/signal-slot": "~1.2",
         "zetacomponents/system-information": "~1.1",
-        "zetacomponents/webdav": "~1.1"
+        "zetacomponents/webdav": "~1.1",
+        "ircmaxell/password-compat": "^1.0"
```
Comment (andrerom) [inline composer.json]: hmm, as we now require php 5.5 as of cf005524485c89f82e6d24e979dfc26d659e45cd, we probably don't need this anymore.

File: settings/site.ini [hunk 4]
```diff
@@ -562,10 +562,12 @@ UserCreatorID=14
 # md5_user generates password hash from user and password.
 # md5_site generates password hash from site, user and password
 # plaintext does not generate a hash but has the password as it is, this is not
-# recommended since it is a huge security risc.
+# recommended since it is a huge security risk.
+# bcrypt generates hash from password only
+# php_default uses whichever method PHP deems appropriate (currently bcrypt, as of PHP 7)
```
Comment (andrerom) [inline settings/site.ini]: `as of PHP 7` => `as of PHP 5.5`: http://php.net/password_hash

File: tests/tests/kernel/datatypes/ezuser/ezuser_test.php [hunk 5]
```diff
@@ -59,9 +59,14 @@ public function testPasswordHashSamePasswordToUser()
...
             $this->fail( "User {$this->username} is not in database.");
         }
-        // Not used in this test
         $passwordHashMD5Password = $rows[0]['password_hash'];
 
+        // The value that is expected to be saved in the ezuser table with the HashType to md5_password
+        $hashMD5Expected = md5( $this->password );
+
+        // Verify that the 2 password hashes saved above are the same
+        $this->assertEquals( $hashMD5Expected, $passwordHashMD5Password );
```
Comment (andrerom) [inline tests/tests/kernel/datatypes/ezuser/ezuser_test.php]: I would kind of expect this to now fail as password in db is blowfish and not md5, rest afterwards I would expect to pass. Strange, no local reproduction of this with mysql?

Comment (glye) [reply tests/tests/kernel/datatypes/ezuser/ezuser_test.php]: All logins here run with `UpdateHash=true`, so at this point the password in db has been changed to `md5_password`. What I don't understand is why it isn't changed to `md5_user` during the next login, which is causing the failing tests. I also have no idea why https://github.com/ezsystems/ezpublish-legacy/pull/1322/commits/7b0136add5f9a06f6e6f819646238a0d3fc15a59 is what triggers it.

Can't reproduce it locally, no.

File: composer.json (hunk 3 above)
Comment (glye) [reply composer.json]: Right! Removed.

File: settings/site.ini (hunk 4 above)
Comment (glye) [reply settings/site.ini]: Fixed, thanks.

**Comment by glye** [comment]
Passing, finally! I'll clean it up and then we're done.

**Comment by emodric** [comment]
@glye So what the root cause of the failure?

**Comment by glye** [comment]
Cleaned up commits, this is how I would like to merge it, when it passes.

@emodric Test data having the same password and the same email address for admin and anon users.

**Comment by emodric** [comment]
> @emodric Test data having the same password and the same email address for admin and anon users.

Huh, so where is it fixed? Can't see anything in the diff.

**Comment by glye** [comment]
@emodric I changed the password for anonymous. Hard to see the change in the hash, since it's changing from MD5 anyway :)

**Comment by emodric** [comment]
> @emodric I changed the password for anonymous. Hard to see the change in the hash, since it's changing from MD5 anyway :)

Hah, okay :) So which one is it now? `anonymous` was before, right?

**Comment by glye** [comment]
It is `anonymous` now, was something else before. Never found out what.

**Comment by emodric** [comment]
:+1:

**Comment by emodric** [comment]
@glye If you wish, you can change the initial commit to give @pedroresende credit instead of me, it's his code to begin with.

**Comment by glye** [comment]
@emodric It's green, not sure I dare to touch it anymore 😁

**Comment by emodric** [comment]
Okay, it's up to you :D

**Comment by glye** [comment]
Re-author-ified now :)

**Comment by emodric** [comment]
@andrerom I think my version was because of `PASSWORD_BCRYPT` => `PASSWORD_HASH_BCRYPT`

**Comment by andrerom** [comment]
@emodric true

**Comment by emodric** [comment]
Yep! Merge away!

**Comment by emodric** [comment]
Hm... Changing the password for a user with old hash types does not update the hash type to the new value.

@andrerom  @glye Do you guys know if this is by design in legacy?

**Pull request merged**